### PR TITLE
Add `cupy.mask_indices`

### PR DESCRIPTION
**Pull request by shwina**
As part of https://github.com/cupy/cupy/issues/6078, adding the `mask_indices` API.

### CODE REVIEW

File: cupy/_sorting/search.py [hunk 1]
```diff
@@ -139,6 +139,41 @@ def nonzero(a):
     _util.check_array(a, arg_name='a')
     return a.nonzero()
 
+def mask_indices(n, mask_func, k=0):
```
Comment (shwina) [inline cupy/_sorting/search.py:142]: Is this the right module for this function? I placed it here because of the similarities to  `nonzero`.

Comment (leofang) [reply cupy/_sorting/search.py:142]: Hi @shwina maybe this is the file to go:
https://github.com/cupy/cupy/blob/c0c5f9e4b7e8693ce6a435276d28e21ae3a55bf5/cupy/_indexing/generate.py#L435

Comment (shwina) [reply cupy/_sorting/search.py:142]: 👍 Thanks!

File: tests/cupy_tests/indexing_tests/test_generate.py [hunk 2]
```diff
@@ -289,3 +289,22 @@ def test_invalid_mode(self, dtype):
             a = tuple([xp.arange(min(dims), dtype=dtype) for d in dims])
             with pytest.raises(ValueError):
                 xp.ravel_multi_index(a, dims, mode='invalid')
+
+
+@testing.gpu
+class TestMaskIndices:
+
+    @testing.numpy_cupy_array_equal()
+    def test_mask_indices(self, xp):
```
Comment (shwina) [inline tests/cupy_tests/indexing_tests/test_generate.py:298]: Note: I tried to parametrize this test over all dtypes, but I believe `cupyx.sparse` (which is  used to generate the data for this test) supports only float and complex types. Is it worth parametrizing over dtypes and generating the input data in a different way?

File: cupy/_indexing/generate.py [hunk 3]
```diff
@@ -426,6 +426,43 @@ def unravel_index(indices, dims, order='C'):
...
+        the locations where ``mask_func(np.ones((n, n)), k)`` is
+        True.
+
+    .. warning::
+
+        This function may synchronize the device.
+
+    .. seealso:: :func:`numpy.mask_indices`
+    """
+    a = cupy.ones((n, n))  # TODO what dtype to use here?
```
Comment (shwina) [inline cupy/_indexing/generate.py:462]: For memory savings, should we use `int8` or `bool` here?

**Comment by takagi** [comment]
Would you also add `mask_indices` to `docs/source/reference/indexing.rst` so that the  document of `mask_indices` is generated?

File: cupy/_indexing/generate.py (hunk 3 above)
Comment (takagi) [reply cupy/_indexing/generate.py:462]: Yes, `int8` suits here.

**Review by takagi — COMMENTED** [review]
Thanks for your pull request! The code is almost fine, and would you check some comments?

File: tests/cupy_tests/indexing_tests/test_generate.py (hunk 2 above)
Comment (shwina) [reply tests/cupy_tests/indexing_tests/test_generate.py:297]: Thanks! I used your idea  of a boolean array as a multiplier to generate a square random matrix of 50% density (exactly  what we were doing before)

**Comment by shwina** [comment]
Thanks for reviewing @takagi! I've addressed your comments. Please let me know if you think this PR needs anything else!

**Review by takagi — APPROVED** [review]

**Comment by takagi** [comment]
Thanks! Let me push a fix to your repo as it is a minor change, and now I've kicked the CI. Your PR is going to be merged soon after CI passes.

**Pull request merged**